### PR TITLE
Improve the naming of exported snapshots. 

### LINKED ISSUE #1755: Improve the naming of exported snapshots.
**Issue summary**

Right now the default snapshot filename looks like this: `forest_snapshot_2022-July-28_height_1165085.car`. We can improve on this by including the network type, using numbers for the months, and allowing custom formatting.

Sub-tasks:
- Include network name (eg. mainnet or calibnet) by default.
- Switch from month names to month numbers (ie. `07` instead of `July`). ISO 8601 date format.
- Allow substitutions in user-specified filenames. For example `-o snapshot_{network}_{date}_{epoch}.car` should result in a `snapshot_calibnet_2022_07_28_1165085.car` file.

**Other information and links**

### PR DESCRIPTION
**Pull request by tyshko5**
**Summary of changes**
Changes introduced in this pull request:
- 

**Reference issue to close (if applicable)**

Closes #1755

**Other information and links**

### CODE REVIEW

File: forest/src/cli/chain_cmd.rs [hunk 1]
```diff
@@ -82,18 +85,31 @@ impl ChainCommands {
...
-                        let now = OffsetDateTime::now_utc();
-                        format!(
-                            "forest_snapshot_{}-{}-{}_height_{}.car",
-                            now.year(),
-                            now.month(),
-                            now.day(),
-                            epoch,
-                        )
-                    }
+                    Some(path) => strfmt(&path, &vars).unwrap(),
```
Comment (elmattic) [inline forest/src/cli/chain_cmd.rs:108]: Is `unwrap` call safe here? If yes some comment would be nice.

File: forest/src/cli/chain_cmd.rs [hunk 2]
```diff
@@ -82,18 +85,31 @@ impl ChainCommands {
...
+                };
+                let year = now.year();
+                let day = now.day();
+                let network_name = cfg.chain.name.clone();
+
+                let mut vars = HashMap::new();
+                vars.insert("year".to_string(), year.to_string());
+                vars.insert("month".to_string(), month_string.clone());
+                vars.insert("day".to_string(), day.to_string());
+                vars.insert("network".to_string(), network_name.clone());
```
Comment (elmattic) [inline forest/src/cli/chain_cmd.rs:104]: Maybe use `chain` instead of `network` to be more coherent with our configuration file.

**Review by elmattic — CHANGES_REQUESTED** [review]

File: forest/src/cli/chain_cmd.rs [hunk 3]
```diff
@@ -82,18 +85,32 @@ impl ChainCommands {
...
                 let epoch = tipset.unwrap_or(chain_head.epoch());
 
+                let now = OffsetDateTime::now_utc();
+
+                let month = now.month() as u8;
+                let month_string = if month < 10 {
+                    format!("0{}", month)
+                } else {
+                    month.to_string()
+                };
```
Comment (LesnyRumcajs) [inline forest/src/cli/chain_cmd.rs:95]: I believe this entire block can be replaced with `let month = format!("{:02}", now.month());`, unless I'm missing something.

File: forest/src/cli/chain_cmd.rs [hunk 4]
```diff
@@ -82,18 +85,32 @@ impl ChainCommands {
...
+                let year = now.year();
+                let day = now.day();
+                let chain_name = cfg.chain.name.clone();
+
+                let mut vars = HashMap::new();
```
Comment (LesnyRumcajs) [inline forest/src/cli/chain_cmd.rs:100]: 1. Do we always need to build this map and pay for it? Or can we do it *only* if we reach `Some(path)` in a `match` later on?
2. That's a lot of cloning, would be nice if we could reduce it. It's not a performance-critical path though, so it's not a big deal.

**Review by LesnyRumcajs — CHANGES_REQUESTED** [review]

**Review by elmattic — CHANGES_REQUESTED** [review]

File: forest/src/cli/chain_cmd.rs (hunk 1 above)
Comment (tyshko5) [reply forest/src/cli/chain_cmd.rs]: Added

File: forest/src/cli/chain_cmd.rs (hunk 3 above)
Comment (tyshko5) [reply forest/src/cli/chain_cmd.rs]: Thanks

File: forest/src/cli/chain_cmd.rs (hunk 4 above)
Comment (tyshko5) [reply forest/src/cli/chain_cmd.rs]: Done both

File: forest/src/cli/chain_cmd.rs [hunk 5]
```diff
@@ -82,18 +85,31 @@ impl ChainCommands {
 
                 let epoch = tipset.unwrap_or(chain_head.epoch());
 
+                let now = OffsetDateTime::now_utc();
+
+                let month = now.month() as u8;
+                let month_string = if month < 10 {
```
[the remark this answers is not in the document]
Comment (tyshko5) [reply forest/src/cli/chain_cmd.rs]: Thanks

File: forest/src/cli/chain_cmd.rs [hunk 6]
```diff
@@ -82,18 +85,28 @@ impl ChainCommands {
...
-                        )
+                    Some(path) => {
+                        let mut vars = HashMap::new();
+                        vars.insert("year".to_string(), year.to_string());
+                        vars.insert("month".to_string(), month_string);
+                        vars.insert("day".to_string(), day.to_string());
+                        vars.insert("chain".to_string(), chain_name);
+                        vars.insert("heigth".to_string(), epoch.to_string());
+                        // we need to unwrap here, as in case the string is malformed one needs to stop the execution
+                        strfmt(path, &vars).unwrap()
```
Comment (lemmih) [inline forest/src/cli/chain_cmd.rs:104]: `unwrap` is not appropriate here. These errors may be shown to the user and should therefore be formatted properly.

File: forest/src/cli/chain_cmd.rs [hunk 7]
```diff
@@ -82,18 +85,28 @@ impl ChainCommands {
...
                     }
+                    None => format!(
+                        "forest_snapshot_{}_{}-{}-{}_height_{}.car",
+                        chain_name, year, month_string, day, epoch,
+                    ),
```
Comment (lemmih) [inline forest/src/cli/chain_cmd.rs:109]: Refactor the code such that we always use `strfmt`. If no path string has been provided, use a default one. You can change the argument parsing to be aware of the default path string.

**Review by lemmih — CHANGES_REQUESTED** [review]

**Review by lemmih — CHANGES_REQUESTED** [review]
Could you change the `output_path` field to be `String` and add `default_value = "forest_snapshot_{chain}_{year}-{month}-{day}_height_{heigth}.car"` to the `structopt` attribute? Also update the help message (run with `--help` and make sure it looks sensible).

File: forest/src/cli/chain_cmd.rs [hunk 8]
```diff
@@ -82,17 +85,32 @@ impl ChainCommands {
...
+
+                let output_path_res = match output_path {
+                    Some(path) => strfmt(path, &vars),
+                    None => strfmt(default_format, &vars),
+                };
+
+                let output_path = match output_path_res {
+                    Ok(path) => path,
+                    Err(e) => {
+                        panic!("Unparsable string error: {}", e);
```
Comment (lemmih) [inline forest/src/cli/chain_cmd.rs:113]: You can't use `panic!` for a user-facing error message. Use `cli_error_and_die`.

**Review by lemmih — CHANGES_REQUESTED** [review]

File: forest/src/cli/chain_cmd.rs [hunk 9]
```diff
@@ -32,9 +38,9 @@ pub enum ChainCommands {
         /// Skip old messages
         #[structopt(short)]
         skip_old_messages: bool,
-        /// Snapshot output path. Default to `forest_snapshot_{year}-{month}-{day}_height_{height}.car`
-        #[structopt(short)]
-        output_path: Option<String>,
+        /// Snapshot output path. Default to `forest_snapshot_{chain_}{year}-{month}-{day}_height_{height}.car`
+        #[structopt(short, default_value = OUTPUT_PATH_DEFAULT_VALUE, help = "format in of the filename of the snapshot")]
```
Comment (lemmih) [inline forest/src/cli/chain_cmd.rs:42]: Use the doc comment as the help string.

**Review by lemmih — CHANGES_REQUESTED** [review]

File: forest/src/cli/chain_cmd.rs [hunk 10]
```diff
@@ -82,17 +88,24 @@ impl ChainCommands {
...
-                            now.month(),
-                            now.day(),
-                            epoch,
-                        )
+                let now = OffsetDateTime::now_utc();
+
+                let month_string = format!("{:02}", now.month() as u8);
+                let year = now.year();
+                let day = now.day();
+                let chain_name = cfg.chain.name.clone();
```
Comment (lemmih) [inline forest/src/cli/chain_cmd.rs:96]: We need to ask the forest node for the chain name. Not sure if we have a RPC call for this or if we need to add one.

File: forest/src/cli/chain_cmd.rs (hunk 9 above)
Comment (lemmih) [reply forest/src/cli/chain_cmd.rs:42]: `StructOpt` can use the doc-comment for the help string. So, you can remove the `help = "..."` part.

**Review by lemmih — CHANGES_REQUESTED** [review]

File: forest/src/cli/chain_cmd.rs (hunk 10 above)
Comment (tyshko5) [reply forest/src/cli/chain_cmd.rs:96]: Should not the chain name be the same as in the configuration?

File: forest/src/cli/chain_cmd.rs (hunk 9 above)
Comment (tyshko5) [reply forest/src/cli/chain_cmd.rs]: Fixed

File: forest/src/cli/chain_cmd.rs (hunk 10 above)
Comment (lemmih) [reply forest/src/cli/chain_cmd.rs:96]: The forest node and the forest RPC can be run with different configurations. They can even be run on different machines.

**Review by lemmih — APPROVED** [review]

File: forest/src/cli/chain_cmd.rs [hunk 11]
```diff
@@ -9,8 +9,13 @@ use crate::cli::{cli_error_and_die, handle_rpc_err};
 use cid::Cid;
 use forest_json::cid::CidJson;
 use rpc_client::chain_ops::*;
+use std::collections::HashMap;
+use strfmt::strfmt;
 use time::OffsetDateTime;
 
+const OUTPUT_PATH_DEFAULT_VALUE: &str =
```
Comment (LesnyRumcajs) [inline forest/src/cli/chain_cmd.rs:16]: I'd say it's more of a `OUTPUT_PATH_DEFAULT_FORMAT`.

File: forest/src/cli/chain_cmd.rs [hunk 12]
```diff
@@ -32,9 +37,9 @@ pub enum ChainCommands {
         /// Skip old messages
         #[structopt(short)]
         skip_old_messages: bool,
-        /// Snapshot output path. Default to `forest_snapshot_{year}-{month}-{day}_height_{height}.car`
-        #[structopt(short)]
-        output_path: Option<String>,
+        /// Snapshot output path. Default to [OUTPUT_PATH_DEFAULT_VALUE]`
```
Comment (LesnyRumcajs) [inline forest/src/cli/chain_cmd.rs:40]: We should put in documentation with what "keywords" the path can be formatted and what they represent (e.g. clearly state whether the time is UTC)

File: forest/src/cli/chain_cmd.rs [hunk 13]
```diff
@@ -82,17 +87,24 @@ impl ChainCommands {
...
+                let year = now.year();
+                let day = now.day();
+                let chain_name = chain_get_name().await.map_err(handle_rpc_err).unwrap();
+
+                let mut vars = HashMap::new();
+                vars.insert("year".to_string(), year.to_string());
+                vars.insert("month".to_string(), month_string);
+                vars.insert("day".to_string(), day.to_string());
+                vars.insert("chain".to_string(), chain_name);
+                vars.insert("heigth".to_string(), epoch.to_string());
```
Comment (LesnyRumcajs) [inline forest/src/cli/chain_cmd.rs:102]: I believe the proper spelling is `height` and not `heigth`

File: forest/src/cli/chain_cmd.rs [hunk 14]
```diff
@@ -9,8 +9,13 @@ use crate::cli::{cli_error_and_die, handle_rpc_err};
 use cid::Cid;
 use forest_json::cid::CidJson;
 use rpc_client::chain_ops::*;
+use std::collections::HashMap;
+use strfmt::strfmt;
 use time::OffsetDateTime;
 
+const OUTPUT_PATH_DEFAULT_VALUE: &str =
+    "forest_snapshot_{chain}_{year}-{month}-{day}_height_{heigth}.car";
```
Comment (LesnyRumcajs) [inline forest/src/cli/chain_cmd.rs:17]: I believe the proper spelling is `height` and not `heigth`

**Review by LesnyRumcajs — CHANGES_REQUESTED** [review]

File: forest/src/cli/chain_cmd.rs (hunk 14 above)
Comment (tyshko5) [reply forest/src/cli/chain_cmd.rs]: Thanks

File: forest/src/cli/chain_cmd.rs (hunk 11 above)
Comment (tyshko5) [reply forest/src/cli/chain_cmd.rs]: Changed

File: forest/src/cli/chain_cmd.rs [hunk 15]
```diff
@@ -32,9 +37,15 @@ pub enum ChainCommands {
...
         #[structopt(short)]
         skip_old_messages: bool,
-        /// Snapshot output path. Default to `forest_snapshot_{year}-{month}-{day}_height_{height}.car`
-        #[structopt(short)]
-        output_path: Option<String>,
+        /// Snapshot output path. Default to [OUTPUT_PATH_DEFAULT_FORMAT]
+        /// Arguments:
+        /// chain - is name of chain name e.g. mainnet 
+        /// year - YYYY format
+        /// maonth - ISO 8601 date format.
```
Comment (lemmih) [inline forest/src/cli/chain_cmd.rs:44]:
```suggestion
        /// month - ISO 8601 date format.
```

**Comment by LesnyRumcajs** [comment]
@tyshko5 please apply `verbatim_doc_comment` to the `output_path` doc. Otherwise, it gets preprocessed and it doesn't look really nice. Currently I get:
```
OPTIONS:
    -o <output-path>
            Snapshot output path. Default to [OUTPUT_PATH_DEFAULT_FORMAT] Arguments: chain - is name of chain name e.g.
            mainnet year - YYYY format month - ISO 8601 date format. day - DD format height - the epoch [default:
            forest_snapshot_{chain}_{year}-{month}-{day}_height_{height}.car]
```

https://docs.rs/structopt/latest/structopt/#doc-comment-preprocessing-and-structoptverbatim_doc_comment

This should preserve new lines and make the doc more readable.

File: forest/src/cli/chain_cmd.rs [hunk 16]
```diff
@@ -72,8 +83,8 @@ impl ChainCommands {
             Self::Export {
                 tipset,
                 recent_stateroots,
-                include_old_messages,
                 output_path,
+                skip_old_messages,
```
Comment (lemmih) [inline forest/src/cli/chain_cmd.rs:87]: This looks like a merge mishap.

**Review by lemmih — APPROVED** [review]

File: forest/src/cli/chain_cmd.rs [hunk 17]
```diff
@@ -82,17 +93,24 @@ impl ChainCommands {
...
-                        )
+                let now = OffsetDateTime::now_utc();
+
+                let month_string = format!("{:02}", now.month() as u8);
+                let year = now.year();
+                let day = now.day();
+                let chain_name = chain_get_name().await.map_err(handle_rpc_err).unwrap();
+
+                let mut vars = HashMap::new();
+                vars.insert("year".to_string(), year.to_string());
```
Comment (LesnyRumcajs) [inline forest/src/cli/chain_cmd.rs:104]: do we need `.to_string()` for keys in the map?

**Review by LesnyRumcajs — CHANGES_REQUESTED** [review]

File: forest/src/cli/chain_cmd.rs [hunk 18]
```diff
@@ -32,9 +37,15 @@ pub enum ChainCommands {
         /// Include old messages
         #[structopt(short, long)]
         include_old_messages: bool,
-        /// Snapshot output path. Default to `forest_snapshot_{year}-{month}-{day}_height_{height}.car`
-        #[structopt(short, long)]
-        output_path: Option<String>,
+        /// Snapshot output path. Default to forest_snapshot_{chain}_{year}-{month}-{day}_height_{height}.car
```
Comment (elmattic) [inline forest/src/cli/chain_cmd.rs:40]: If we keep `month` in letters and not in `YYYY-MM-DD` format, I would put `_height_{height}` before in the default output path so they are displayed nicely when sorted alphabetical order (ie on DO Spaces).

What about support for `HH-MM-SS` if user want it?

Comment (elmattic) [inline forest/src/cli/chain_cmd.rs:40]: In the doc comment you can add `` around `forest_snapshot_{chain}_{year}-{month}-{day}_height_{height}.car` for better markdown display.

**Review by elmattic — CHANGES_REQUESTED** [review]

File: forest/src/cli/chain_cmd.rs [hunk 19]
```diff
@@ -82,17 +93,24 @@ impl ChainCommands {
...
-                        format!(
-                            "forest_snapshot_{}-{}-{}_height_{}.car",
-                            now.year(),
-                            now.month(),
-                            now.day(),
-                            epoch,
-                        )
+                let now = OffsetDateTime::now_utc();
+
+                let month_string = format!("{:02}", now.month() as u8);
```
Comment (elmattic) [inline forest/src/cli/chain_cmd.rs:98]: Why do we have this zero padding for `month` and not `day`?

File: forest/src/cli/chain_cmd.rs (hunk 18 above)
Comment (lemmih) [reply forest/src/cli/chain_cmd.rs:40]: I think it's an oversight that the month is not numerical. Should follow the ISO 8601 date format by default.

File: forest/src/cli/chain_cmd.rs (hunk 17 above)
Comment (tyshko5) [reply forest/src/cli/chain_cmd.rs:104]: Yes, we do.

File: forest/src/cli/chain_cmd.rs (hunk 18 above)
Comment (tyshko5) [reply forest/src/cli/chain_cmd.rs]: Done

File: forest/src/cli/chain_cmd.rs [hunk 20]
```diff
@@ -82,17 +93,24 @@ impl ChainCommands {
...
+                let year = now.year();
+                let day = now.day();
+                let chain_name = chain_get_name().await.map_err(handle_rpc_err).unwrap();
+
+                let mut vars = HashMap::new();
```
[the remark this answers is not in the document]
Comment (tyshko5) [reply forest/src/cli/chain_cmd.rs]: Done

**Review by LesnyRumcajs — APPROVED** [review]

**Review by elmattic — APPROVED** [review]

**Pull request merged**